lib/persisted-model.js

```diff
@@ -651,6 +673,22 @@ module.exports = function(registry) {
 
     setRemoting(PersistedModel, 'replaceOrCreate', replaceOrCreateOptions);
 
+    setRemoting(PersistedModel, 'upsertWithWhere', {
+      aliases: ['patchOrCreateWithWhere'],
+      description: 'Update an existing model instance or ' +
+      'insert a new one into the data source based on the ' +
+      'where criteria.',
+      accessType: 'WRITE',
+      accepts: [
+        { arg: 'where', type: 'object', http: { source: 'query' },
+          description: 'Criteria to match model instances' },
+        { arg: 'data', type: 'object', http: { source: 'body' },
+          description: 'An object of model property name/value pairs' },
+      ],
+      returns: { arg: 'data', type: typeName, root: true },
+      http: { verb: 'put', path: '/upsertWithWhere' },
```

Comment:
>  Could you please confirm me that since this is a new method which is being introduced we don't need to make it configurable like the way we did with replaceOnPUT flag for replaceOrCreate/updateOrCreate and replaceById/patchAttributes, right?

Yes, no need to make this configurable.

> whether they should go for POST /updateOrCreate or PATCH /updateOrCreate? I have no strong opinion in this regard.

Let's go with `POST /upsertWithWhere` or even `POST /updateOrCreateWithWhere`.